contracts/BucketAuction.sol

```diff
@@ -0,0 +1,282 @@
+        _auctionActive = b;
+    }
+
+    /**
+     * @notice place a bid in ETH or add to your existing bid. Calling this
+     *   multiple times will increase your bid amount. All bids placed are final
+     *   and cannot be reversed.
+     */
+    function bid() external payable isAuctionActive nonReentrant cannotMint {
+        User storage bidder = _userData[msg.sender]; // get user's current bid total
+        uint256 contribution_ = bidder.contribution; // bidder.contribution is uint216
+        unchecked {
+            // does not overflow
+            contribution_ += msg.value;
+        }
+        if (contribution_ < _minimumContributionInWei)
+            revert LowerThanMinBidAmount();
+        bidder.contribution = uint216(contribution_);
+        emit Bid(msg.sender, msg.value, contribution_, address(this).balance);
+    }
+
+    /**
+     * @notice set the minimum contribution required to place a bid
+     * @dev set this price in wei, not eth!
+     * @param minimumContributionInWei new price, set in wei
+     */
+    function setMinimumContribution(uint256 minimumContributionInWei)
+        external
+        onlyOwner
+    {
+        _minimumContributionInWei = minimumContributionInWei;
+        emit SetMinimumContribution(minimumContributionInWei);
+    }
+
+    /**
+     * @notice set the clearing price after all bids have been placed.
+     * @dev set this price in wei, not eth!
+     * @param priceInWei new price, set in wei
+     */
+    function setPrice(uint256 priceInWei)
+        external
+        isAuctionInactive
+        onlyOwner
+        cannotMint
+    {
+        _price = priceInWei;
+        emit SetPrice(priceInWei);
+    }
+
+    /**
+     * @dev handles all minting.
+     * @param to address to mint tokens to.
+     * @param numberOfTokens number of tokens to mint.
+     */
+    function _internalMint(address to, uint256 numberOfTokens)
+        internal
+        hasSupply(numberOfTokens)
+    {
+        _safeMint(to, numberOfTokens);
+    }
+
+    /**
+     * @dev handles multiple send tokens methods.
+     * @param to address to send tokens to.
+     * @param n number of tokens to send.
+     */
+    function _sendTokens(address to, uint256 n) internal {
+        uint256 price = _price; // storage to memory
+        if (price == 0) revert PriceNotSet();
+
+        User storage user = _userData[to]; // get user data
+        uint256 claimed = user.tokensClaimed; // user.tokensClaimed is uint32
+        claimed += n;
+
+        if (claimed > (user.contribution / price))
+            revert CannotSendMoreThanUserPurchased();
+        user.tokensClaimed = uint32(claimed);
+        _internalMint(to, n);
+    }
+
+    /**
+     * @notice get the number of tokens purchased by an address, after the
+     *   clearing price has been set.
+     * @param a address to query.
+     */
+    function amountPurchased(address a) public view returns (uint256) {
+        return _userData[a].contribution / _price;
```

Comment:
qq: the # tokens are decided by `contribution` and `_price` here. But do we need to support the case that user explicitly bid for X tokens with Y contribution?

e.g. I bid 1 token for 200 wei. The price is set to 100. I just get 1 token and 100 wei refund.

Reply:
no, that's not bucket auction - that's just auction with multiple bids